cpp/include/raft/core/math.hpp

```diff
@@ -91,7 +96,14 @@ RAFT_INLINE_FUNCTION auto atanh(T x)
 
 /** Cosine */
 template <typename T>
-RAFT_INLINE_FUNCTION auto cos(T x)
+RAFT_INLINE_FUNCTION
+#if _RAFT_HAS_CUDA
+  typename std::enable_if<!std::is_same<T, __half>::value && !std::is_same<T, nv_bfloat16>::value,
+                          T>::type
+#else
+  auto
+#endif
```

Comment:
Can you use `CUDA_ONLY_CONDITION` / `CUDA_CONDITION_ELSE_TRUE` when you have this kind of host guard?